agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/utils/ResourceParser.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.azure.monitor.opentelemetry.exporter.implementation.utils;
+
+import com.azure.monitor.opentelemetry.exporter.implementation.builders.AbstractTelemetryBuilder;
+import com.azure.monitor.opentelemetry.exporter.implementation.models.ContextTagKeys;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public final class ResourceParser {
+
+  public static void updateRoleNameAndInstance(
+      AbstractTelemetryBuilder builder, Resource resource) {
+    Map<ContextTagKeys, String> resourceMap = ResourceParser.parseRoleNameAndInstance(resource);
+    if (resourceMap != null && !resourceMap.isEmpty()) {
+      resourceMap.forEach(
+          (key, value) -> {
+            String stringKey = key.toString();
+            if (ContextTagKeys.AI_CLOUD_ROLE.toString().equals(stringKey)
+                || ContextTagKeys.AI_CLOUD_ROLE_INSTANCE.toString().equals(stringKey)) {
+              builder.addTag(stringKey, value);
+            }
+          });
+    }
+  }
+
+  // visible for test
+  @Nullable
+  static Map<ContextTagKeys, String> parseRoleNameAndInstance(@Nullable Resource resource) {
```

Comment:
can we call this once at startup and cache the resulting map? Resources aren't allowed to change (currently at least) in OpenTelemetry

Reply:
no, resource per metric data

Reply:
ya, but resources are immutable, tied to the `OpenTelemetry` instance, and shared across all telemetry.

it's possible for users to create multiple `OpenTelemetry` instances though, so that would be a reason not to cache.

maybe pass in the telemetry builder, so we don't need to create the extra Map on each call (just as an optimization to reduce memory allocation)

Reply:
passing in builder.